jax/api.py

```diff
@@ -297,22 +296,77 @@ def cache_miss(*args, **kw):
     options = xb.get_compile_options(
         num_replicas=1,
         num_partitions=1,
-        device_assignment=(device.id,) if device else None)
+        device_assignment=(device_.id,) if device_ else None)
     if xla_result.parameter_is_tupled_arguments:
       options.parameter_is_tupled_arguments = True
 
     return (
         # We call xla.backend_compile to ensure the XLA computation appears
         # separately in Python profiling results.
-        xla.backend_compile(backend, xla_result.xla_computation, options),
+        xla.backend_compile(backend_, xla_result.xla_computation, options),
         xla_result.out_pytree_def,
         xla_result.shaped_arrays,
         xla_result.lazy_expressions)
 
   # Delay the import, because it requires a new version of jaxlib.
   from .lib import jax_jit  # pylint: disable=g-import-not-at-top
-  return jax_jit.jit(cache_miss, python_jitted_f, FLAGS.jax_enable_x64,
-                     static_argnums, device)
+  cpp_jitted_f = jax_jit.jit(fun, cache_miss,
+                             python_jitted_f, FLAGS.jax_enable_x64,
+                             config.read("jax_disable_jit"), static_argnums)
+
+  # The following exists for the ony purpose of supporting the following usage:
+  # class A:
+  #   @functools.partial(jax.jit, static_argnums=0)
+  #   def _compute_log_data(self, ...)
+  #     ...
+  # We discourage this practice as it's not a pure stateless functions, but
+  # the C++ codepath supports this for backward compatibility.
+  # Note that the JAX team is thinking of requiring the static arguments to be
+  # hashable, so maybe such usage will be forbidden in the future.
+  # Prefer for example an attribute or property access:
+  # class A:
+  #   def __init__(self):
+  #     self.jitted_f = jax.jit(self.f)
+
+  # We try to do this more expensive path only when it's needed and assume
+  # users respect the `self` convention.
+  # See `_IsMethodDecorator`
+  # Note that @classmetod and @staticmethod are *not* supported.
+  if 0 in static_argnums and inspect.getfullargspec(fun).args[0] == "self":
+    return _IsMethodDecorator(cpp_jitted_f)
+  else:
+    return cpp_jitted_f
+
+
+class _IsMethodDecorator:
+  """Detects whether `fun` is a method and resolves the instance.
+
+  A function is decorated *before* it is bound, so we cannot directly
+  determine whether it is a 'method' or 'function'. We can rely on the
+  Descriptor protocol (see below) to detect, at *call time* whether it was a
+  method and retrieve the associated object.
+
+  We prefer to do this only when we think it' needed as the goal is to remove
+  all overheads.
+
+  https://docs.python.org/3/howto/descriptor.html#functions-and-methods
+  """
+
+  def __init__(self, func):
+    self.func = func
+    self.is_method = False
+
+  def __get__(self, instance, owner):
+    if not self.is_method:
```

Comment:
Is it important to have this conditional?

Reply:
I guess not, removed.